compiler/rustc_lint/src/lifetime_style.rs

```diff
@@ -0,0 +1,328 @@
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{self as hir};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::Span;
+use tracing::instrument;
+
+use crate::lints::{self, MismatchedElidedLifetimeStylesSuggestion};
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `mismatched_elided_lifetime_styles` lint detects when an
+    /// elided lifetime uses different syntax between function
+    /// arguments and return values.
+    ///
+    /// The three kinds of syntax are:
+    /// 1. Named lifetimes, such as `'static` or `'a`.
+    /// 2. The anonymous lifetime `'_`.
+    /// 3. Hidden lifetimes, such as `&u8` or `ThisHasALifetimeGeneric`.
+    ///
+    /// As an exception, this lint allows references with hidden
+    /// lifetimes to be paired with paths using anonymous lifetimes.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(mismatched_elided_lifetime_styles)]
+    /// pub fn mixing_a_name_with_hidden(v: &'static u8) -> &u8 {
+    ///     v
+    /// }
+    ///
+    /// pub fn mixing_hidden_with_anonymous(v: &u8) -> &'_ u8 {
+    ///     v
+    /// }
+    ///
+    /// struct Foo;
+    /// impl Foo {
+    ///     // Lifetime elision results in the output lifetime becoming
+    ///     // `'static`, which is not what was intended.
+    ///     pub fn get_mut(&'static self, x: &mut u8) -> &mut u8 {
+    ///         unsafe { &mut *(x as *mut _) }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Lifetime elision is useful because it frees you from having to
+    /// give each lifetime its own name and explicitly show the
+    /// relation of input and output lifetimes for common
+    /// cases. However, a lifetime that uses inconsistent syntax
+    /// between the arguments and return values is more confusing.
+    ///
+    /// In certain `unsafe` code, lifetime elision combined with
+    /// inconsistent lifetime syntax may result in unsound code.
+    pub MISMATCHED_ELIDED_LIFETIME_STYLES,
```

Comment:
```suggestion
    pub MISMATCHED_LIFETIME_STYLES,
```

I'd suggest calling this `mismatched_lifetime_styles`.  There are in my view 1) named lifetimes, 2) elided lifetimes, and 3) hidden lifetimes.

Therefore, this is not a mismatch of different kinds of elided lifetimes, it's a mismatch of different kinds of "lifetimes" (i.e. generic lifetime parameter syntaxes).

In particular, `&'a ()` is distinctly not an elided lifetime, so it can't be part of a mismatch between elided lifetime styles.